```
Error running mint automation
+ mkdir -p 16146-95552af/gopath/src/github.com/minio
+ git -C 16146-95552af/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 16146-95552af/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 16146-95552af/gopath/src/github.com/minio/minio fetch --quiet minio pull/16146/head:pr16146
+ git -C 16146-95552af/gopath/src/github.com/minio/minio checkout --quiet pr16146
+ GO111MODULE=on
+ GOPATH=/home/mint-auto/mint-auto/16146-95552af/gopath
+ make -C 16146-95552af/gopath/src/github.com/minio/minio --quiet
go: downloading github.com/cosnicolaou/pbzip2 v1.0.1
go: downloading cloud.google.com/go/storage v1.26.0
go: downloading github.com/minio/pkg v1.5.4
go: downloading github.com/minio/console v0.21.3
go: downloading github.com/minio/selfupdate v0.5.0
go: downloading github.com/minio/xxml v0.0.3
go: downloading github.com/minio/dperf v0.4.2
go: downloading github.com/inconshreveable/mousetrap v1.0.1
go: downloading github.com/djherbis/atime v1.1.0
go: downloading github.com/hashicorp/golang-lru v0.5.4
go: downloading github.com/go-openapi/loads v0.21.2
go: downloading github.com/felixge/fgprof v0.9.3
go: downloading github.com/minio/cli v1.24.0
go: downloading github.com/golang-jwt/jwt/v4 v4.4.2
go: downloading github.com/klauspost/compress v1.15.11
go: downloading github.com/minio/minio-go/v7 v7.0.43
go: downloading github.com/minio/zipindex v0.3.0
go: downloading github.com/lithammer/shortuuid/v4 v4.0.0
go: downloading github.com/miekg/dns v1.1.50
go: downloading github.com/json-iterator/go v1.1.12
go: downloading github.com/klauspost/pgzip v1.2.5
go: downloading github.com/klauspost/readahead v1.4.0
go: downloading github.com/minio/highwayhash v1.0.2
go: downloading github.com/mitchellh/go-homedir v1.1.0
go: downloading github.com/cespare/xxhash/v2 v2.1.2
go: downloading github.com/fatih/color v1.13.0
go: downloading github.com/dustin/go-humanize v1.0.0
go: downloading github.com/prometheus/client_model v0.3.0
go: downloading golang.org/x/crypto v0.2.0
go: downloading github.com/klauspost/reedsolomon v1.11.0
go: downloading github.com/google/uuid v1.3.0
go: downloading github.com/prometheus/common v0.37.0
go: downloading github.com/gorilla/mux v1.8.0
go: downloading github.com/cheggaaa/pb v1.0.29
go: downloading github.com/minio/kes v0.22.0
go: downloading github.com/prometheus/procfs v0.8.0
go: downloading github.com/bits-and-blooms/bloom/v3 v3.3.1
go: downloading github.com/dchest/siphash v1.2.3
go: downloading github.com/Azure/azure-storage-blob-go v0.15.0
go: downloading github.com/bcicen/jstream v1.0.1
go: downloading github.com/coreos/go-oidc v2.2.1+incompatible
go: downloading golang.org/x/time v0.0.0-20220722155302-e5dcc9cfc0b9
go: downloading github.com/beevik/ntp v0.3.0
go: downloading github.com/coredns/coredns v1.9.4
go: downloading github.com/go-ldap/ldap/v3 v3.4.4
go: downloading go.etcd.io/etcd/client/v3 v3.5.5
go: downloading github.com/ncw/directio v1.0.5
go: downloading go.uber.org/zap v1.23.0
go: downloading github.com/tinylib/msgp v1.1.7-0.20220719154719-f3635b96e483
go: downloading golang.org/x/sys v0.2.0
go: downloading github.com/minio/sio v0.3.0
go: downloading github.com/minio/sha256-simd v1.0.0
go: downloading github.com/philhofer/fwd v1.1.2-0.20210722190033-5c56ac6d0bb9
go: downloading github.com/pierrec/lz4 v2.6.1+incompatible
go: downloading github.com/prometheus/client_golang v1.13.0
go: downloading github.com/rs/cors v1.8.2
go: downloading github.com/rs/dnscache v0.0.0-20211102005908-e0241e321417
go: downloading github.com/secure-io/sio-go v0.3.1
go: downloading github.com/shirou/gopsutil/v3 v3.22.9
go: downloading github.com/valyala/bytebufferpool v1.0.0
go: downloading github.com/yargevad/filepathx v1.0.0
go: downloading github.com/zeebo/xxh3 v1.0.2
go: downloading go.etcd.io/etcd/api/v3 v3.5.5
go: downloading go.uber.org/atomic v1.10.0
go: downloading golang.org/x/oauth2 v0.1.0
go: downloading google.golang.org/api v0.98.0
go: downloading gopkg.in/yaml.v2 v2.4.0
go: downloading github.com/buger/jsonparser v1.1.1
go: downloading github.com/Shopify/sarama v1.36.0
go: downloading github.com/eclipse/paho.mqtt.golang v1.4.1
go: downloading github.com/elastic/go-elasticsearch/v7 v7.17.7
go: downloading github.com/go-sql-driver/mysql v1.6.0
go: downloading github.com/gomodule/redigo v1.8.9
go: downloading github.com/nats-io/stan.go v0.10.3
go: downloading github.com/lib/pq v1.10.7
go: downloading github.com/nats-io/nats.go v1.17.0
go: downloading github.com/nsqio/go-nsq v1.1.0
go: downloading github.com/pkg/errors v0.9.1
go: downloading github.com/rabbitmq/amqp091-go v1.5.0
go: downloading github.com/xdg/scram v1.0.5
go: downloading github.com/minio/simdjson-go v0.4.2
go: downloading github.com/golang/protobuf v1.5.2
go: downloading github.com/mattn/go-colorable v0.1.13
go: downloading github.com/mattn/go-isatty v0.0.16
go: downloading golang.org/x/sync v0.1.0
go: downloading github.com/bits-and-blooms/bitset v1.3.3
go: downloading golang.org/x/net v0.2.0
go: downloading github.com/mattn/go-runewidth v0.0.14
go: downloading github.com/pquerna/cachecontrol v0.1.0
go: downloading gopkg.in/square/go-jose.v2 v2.6.0
go: downloading github.com/klauspost/cpuid/v2 v2.1.2
go: downloading github.com/minio/csvparser v1.0.0
go: downloading github.com/fraugster/parquet-go v0.12.0
go: downloading github.com/alecthomas/participle v0.7.1
go: downloading github.com/Azure/go-ntlmssp v0.0.0-20220621081337-cb9428e4ac1e
go: downloading github.com/go-asn1-ber/asn1-ber v1.5.4
go: downloading go.etcd.io/etcd/client/pkg/v3 v3.5.5
go: downloading google.golang.org/grpc v1.50.1
go: downloading github.com/xdg/stringprep v1.0.3
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.2
go: downloading github.com/golang/snappy v0.0.4
go: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: downloading github.com/modern-go/reflect2 v1.0.2
go: downloading github.com/gogo/protobuf v1.3.2
go: downloading google.golang.org/genproto v0.0.0-20221018160656-63c7b68cfc55
go: downloading github.com/coreos/go-semver v0.3.0
go: downloading github.com/rivo/uniseg v0.4.2
go: downloading github.com/nats-io/nuid v1.0.1
go: downloading go.uber.org/multierr v1.8.0
go: downloading github.com/go-openapi/analysis v0.21.4
go: downloading github.com/go-openapi/spec v0.20.7
go: downloading github.com/go-openapi/swag v0.22.3
go: downloading github.com/Azure/azure-pipeline-go v0.2.3
go: downloading github.com/google/pprof v0.0.0-20220829040838-70bd9ae97f40
go: downloading github.com/lestrrat-go/jwx v1.2.25
go: downloading github.com/rjeczalik/notify v0.9.2
go: downloading github.com/fatih/structs v1.1.0
go: downloading github.com/montanaflynn/stats v0.6.6
go: downloading github.com/gorilla/websocket v1.5.0
go: downloading github.com/beorn7/perks v1.0.1
go: downloading google.golang.org/protobuf v1.28.1
go: downloading aead.dev/mem v0.2.0
go: downloading golang.org/x/text v0.4.0
go: downloading github.com/mattn/go-ieproxy v0.0.1
go: downloading github.com/rs/xid v1.4.0
go: downloading gopkg.in/ini.v1 v1.67.0
go: downloading github.com/minio/md5-simd v1.1.2
go: downloading gopkg.in/yaml.v3 v3.0.1
go: downloading github.com/mailru/easyjson v0.7.7
go: downloading github.com/coreos/go-systemd/v22 v22.4.0
go: downloading cloud.google.com/go v0.104.0
go: downloading github.com/nats-io/nkeys v0.3.0
go: downloading github.com/josharian/intern v1.0.0
go: downloading github.com/apache/thrift v0.16.0
go: downloading github.com/eapache/go-resiliency v1.3.0
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading github.com/jcmturner/gokrb5/v8 v8.4.3
go: downloading github.com/eapache/go-xerial-snappy v0.0.0-20180814174437-776d5712da21
go: downloading github.com/rcrowley/go-metrics v0.0.0-20201227073835-cf1acfcdf475
go: downloading github.com/jcmturner/gofork v1.7.6
go: downloading github.com/eapache/queue v1.1.0
go: downloading github.com/hashicorp/go-multierror v1.1.1
go: downloading github.com/pierrec/lz4/v4 v4.1.16
go: downloading github.com/lestrrat-go/iter v1.0.2
go: downloading github.com/lestrrat-go/backoff/v2 v2.0.8
go: downloading github.com/lestrrat-go/option v1.0.0
go: downloading github.com/lestrrat-go/httpcc v1.0.1
go: downloading github.com/lestrrat-go/blackmagic v1.0.1
go: downloading github.com/go-openapi/strfmt v0.21.3
go: downloading github.com/go-openapi/jsonpointer v0.19.5
go: downloading github.com/go-openapi/jsonreference v0.20.0
go: downloading github.com/hashicorp/errwrap v1.1.0
go: downloading cloud.google.com/go/iam v0.4.0
go: downloading cloud.google.com/go/compute v1.10.0
go: downloading github.com/googleapis/gax-go/v2 v2.5.1
go: downloading go.opencensus.io v0.23.0
go: downloading golang.org/x/xerrors v0.0.0-20220907171357-04be3eba64a2
go: downloading github.com/go-openapi/errors v0.20.3
go: downloading github.com/oklog/ulid v1.3.1
go: downloading go.mongodb.org/mongo-driver v1.10.3
go: downloading github.com/mitchellh/mapstructure v1.5.0
go: downloading github.com/asaskevich/govalidator v0.0.0-20210307081110-f21760c49a8d
go: downloading github.com/hashicorp/go-uuid v1.0.3
go: downloading github.com/jcmturner/dnsutils/v2 v2.0.0
go: downloading github.com/jcmturner/aescts/v2 v2.0.0
go: downloading github.com/jcmturner/rpc/v2 v2.0.3
go: downloading github.com/golang/groupcache v0.0.0-20210331224755-41bb18bfe9da
go: downloading github.com/go-openapi/runtime v0.24.2
go: downloading github.com/go-openapi/validate v0.22.0
go: downloading github.com/jessevdk/go-flags v1.5.0
go: downloading github.com/minio/mc v0.0.0-20221103000258-583d449e38cd
go: downloading github.com/unrolled/secure v1.13.0
go: downloading github.com/tidwall/gjson v1.14.3
go: downloading github.com/tidwall/match v1.1.1
go: downloading github.com/tidwall/pretty v1.2.1
go: downloading github.com/docker/go-units v0.5.0
go: downloading github.com/charmbracelet/bubbles v0.14.0
go: downloading github.com/google/shlex v0.0.0-20191202100458-e7afc7fbc510
go: downloading github.com/navidys/tvxwidgets v0.1.1
go: downloading gopkg.in/h2non/filetype.v1 v1.0.5
go: downloading github.com/posener/complete v1.2.3
go: downloading github.com/charmbracelet/lipgloss v0.6.0
go: downloading golang.org/x/term v0.2.0
go: downloading github.com/gdamore/tcell/v2 v2.5.3
go: downloading github.com/minio/colorjson v1.0.4
go: downloading github.com/pkg/xattr v0.4.9
go: downloading github.com/olekukonko/tablewriter v0.0.5
go: downloading github.com/muesli/reflow v0.3.0
go: downloading github.com/prometheus/prom2json v1.3.2
go: downloading github.com/charmbracelet/bubbletea v0.22.1
go: downloading github.com/rivo/tview v0.0.0-20220916081518-2e69b7385a37
go: downloading github.com/minio/filepath v1.0.0
go: downloading github.com/tklauser/go-sysconf v0.3.11
go: downloading github.com/muesli/termenv v0.13.0
go: downloading github.com/muesli/cancelreader v0.2.2
go: downloading github.com/containerd/console v1.0.3
go: downloading github.com/mattn/go-localereader v0.0.1
go: downloading github.com/muesli/ansi v0.0.0-20211031195517-c9f0611b6c70
go: downloading github.com/tklauser/numcpus v0.6.0
go: downloading github.com/lucasb-eyer/go-colorful v1.2.0
go: downloading github.com/aymanbagabas/go-osc52 v1.2.1
go: downloading github.com/gdamore/encoding v1.0.0
go: downloading github.com/googleapis/enterprise-certificate-proxy v0.1.0
go: downloading github.com/google/go-cmp v0.5.9
../../../../pkg/mod/github.com/minio/console@v0.21.3/pkg/logger/config/config.go:20:2: github.com/minio/madmin-go@v1.7.5: replacement directory ../madmin-go does not exist
../../../../pkg/mod/github.com/minio/pkg@v1.5.4/sys/stats_linux.go:27:2: github.com/minio/madmin-go@v1.7.5: replacement directory ../madmin-go does not exist
cmd/admin-handlers.go:51:2: github.com/minio/madmin-go@v1.7.5: replacement directory ../madmin-go does not exist
cmd/server-rlimit.go:24:2: github.com/minio/madmin-go@v1.7.5: replacement directory ../madmin-go does not exist
make: *** [build] Error 1

```

--- a/cmd/auth-handler.go
+++ b/cmd/auth-handler.go
@@ -39,6 +39,7 @@ import (
 	xhttp "github.com/minio/minio/internal/http"
 	xjwt "github.com/minio/minio/internal/jwt"
 	"github.com/minio/minio/internal/logger"
+	"github.com/minio/minio/internal/mcontext"
 	"github.com/minio/pkg/bucket/policy"
 	iampolicy "github.com/minio/pkg/iam/policy"
 )
@@ -544,7 +545,7 @@ func isSupportedS3AuthType(aType authType) bool {
 func setAuthHandler(h http.Handler) http.Handler {
 	// handler for validating incoming authorization headers.
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		tc, ok := r.Context().Value(contextTraceReqKey).(*traceCtxt)
+		tc, ok := r.Context().Value(mcontext.ContextTraceKey).(*mcontext.TraceCtxt)
 
 		aType := getRequestAuthType(r)
 		if aType == authTypeSigned || aType == authTypeSignedV2 || aType == authTypeStreamingSigned {
@@ -552,8 +553,8 @@ func setAuthHandler(h http.Handler) http.Handler {
 			amzDate, errCode := parseAmzDateHeader(r)
 			if errCode != ErrNone {
 				if ok {
-					tc.funcName = "handler.Auth"
-					tc.responseRecorder.LogErrBody = true
+					tc.FuncName = "handler.Auth"
+					tc.ResponseRecorder.LogErrBody = true
 				}
 
 				// All our internal APIs are sensitive towards Date
@@ -568,8 +569,8 @@ func setAuthHandler(h http.Handler) http.Handler {
 			curTime := UTCNow()
 			if curTime.Sub(amzDate) > globalMaxSkewTime || amzDate.Sub(curTime) > globalMaxSkewTime {
 				if ok {
-					tc.funcName = "handler.Auth"
-					tc.responseRecorder.LogErrBody = true
+					tc.FuncName = "handler.Auth"
+					tc.ResponseRecorder.LogErrBody = true
 				}
 
 				writeErrorResponse(r.Context(), w, errorCodes.ToAPIErr(ErrRequestTimeTooSkewed), r.URL)
@@ -583,8 +584,8 @@ func setAuthHandler(h http.Handler) http.Handler {
 		}
 
 		if ok {
-			tc.funcName = "handler.Auth"
-			tc.responseRecorder.LogErrBody = true
+			tc.FuncName = "handler.Auth"
+			tc.ResponseRecorder.LogErrBody = true
 		}
 
 		writeErrorResponse(r.Context(), w, errorCodes.ToAPIErr(ErrSignatureVersionNotSupported), r.URL)

--- a/cmd/generic-handlers.go
+++ b/cmd/generic-handlers.go
@@ -37,6 +37,7 @@ import (
 	xhttp "github.com/minio/minio/internal/http"
 	"github.com/minio/minio/internal/http/stats"
 	"github.com/minio/minio/internal/logger"
+	"github.com/minio/minio/internal/mcontext"
 )
 
 const (
@@ -101,13 +102,13 @@ func isHTTPHeaderSizeTooLarge(header http.Header) bool {
 // Limits body and header to specific allowed maximum limits as per S3/MinIO API requirements.
 func setRequestLimitHandler(h http.Handler) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		tc, ok := r.Context().Value(contextTraceReqKey).(*traceCtxt)
+		tc, ok := r.Context().Value(mcontext.ContextTraceKey).(*mcontext.TraceCtxt)
 
 		// Reject unsupported reserved metadata first before validation.
 		if containsReservedMetadata(r.Header) {
 			if ok {
-				tc.funcName = "handler.ValidRequest"
-				tc.responseRecorder.LogErrBody = true
+				tc.FuncName = "handler.ValidRequest"
+				tc.ResponseRecorder.LogErrBody = true
 			}
 
 			writeErrorResponse(r.Context(), w, errorCodes.ToAPIErr(ErrUnsupportedMetadata), r.URL)
@@ -116,8 +117,8 @@ func setRequestLimitHandler(h http.Handler) http.Handler {
 
 		if isHTTPHeaderSizeTooLarge(r.Header) {
 			if ok {
-				tc.funcName = "handler.ValidRequest"
-				tc.responseRecorder.LogErrBody = true
+				tc.FuncName = "handler.ValidRequest"
+				tc.ResponseRecorder.LogErrBody = true
 			}
 
 			writeErrorResponse(r.Context(), w, errorCodes.ToAPIErr(ErrMetadataTooLarge), r.URL)
@@ -378,12 +379,12 @@ func hasMultipleAuth(r *http.Request) bool {
 // any malicious requests.
 func setRequestValidityHandler(h http.Handler) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		tc, ok := r.Context().Value(contextTraceReqKey).(*traceCtxt)
+		tc, ok := r.Context().Value(mcontext.ContextTraceKey).(*mcontext.TraceCtxt)
 
 		if err := hasBadHost(r.Host); err != nil {
 			if ok {
-				tc.funcName = "handler.ValidRequest"
-				tc.responseRecorder.LogErrBody = true
+				tc.FuncName = "handler.ValidRequest"
+				tc.ResponseRecorder.LogErrBody = true
 			}
 
 			invalidReq := errorCodes.ToAPIErr(ErrInvalidRequest)
@@ -396,8 +397,8 @@ func setRequestValidityHandler(h http.Handler) http.Handler {
 		// Check for bad components in URL path.
 		if hasBadPathComponent(r.URL.Path) {
 			if ok {
-				tc.funcName = "handler.ValidRequest"
-				tc.responseRecorder.LogErrBody = true
+				tc.FuncName = "handler.ValidRequest"
+				tc.ResponseRecorder.LogErrBody = true
 			}
 
 			writeErrorResponse(r.Context(), w, errorCodes.ToAPIErr(ErrInvalidResourceName), r.URL)
@@ -409,8 +410,8 @@ func setRequestValidityHandler(h http.Handler) http.Handler {
 			for _, v := range vv {
 				if hasBadPathComponent(v) {
 					if ok {
-						tc.funcName = "handler.ValidRequest"
-						tc.responseRecorder.LogErrBody = true
+						tc.FuncName = "handler.ValidRequest"
+						tc.ResponseRecorder.LogErrBody = true
 					}
 
 					writeErrorResponse(r.Context(), w, errorCodes.ToAPIErr(ErrInvalidResourceName), r.URL)
@@ -421,8 +422,8 @@ func setRequestValidityHandler(h http.Handler) http.Handler {
 		}
 		if hasMultipleAuth(r) {
 			if ok {
-				tc.funcName = "handler.Auth"
-				tc.responseRecorder.LogErrBody = true
+				tc.FuncName = "handler.Auth"
+				tc.ResponseRecorder.LogErrBody = true
 			}
 
 			invalidReq := errorCodes.ToAPIErr(ErrInvalidRequest)
@@ -436,8 +437,8 @@ func setRequestValidityHandler(h http.Handler) http.Handler {
 		if isMinioReservedBucket(bucketName) || isMinioMetaBucket(bucketName) {
 			if !guessIsRPCReq(r) && !guessIsBrowserReq(r) && !guessIsHealthCheckReq(r) && !guessIsMetricsReq(r) && !isAdminReq(r) && !isKMSReq(r) {
 				if ok {
-					tc.funcName = "handler.ValidRequest"
-					tc.responseRecorder.LogErrBody = true
+					tc.FuncName = "handler.ValidRequest"
+					tc.ResponseRecorder.LogErrBody = true
 				}
 				writeErrorResponse(r.Context(), w, errorCodes.ToAPIErr(ErrAllAccessDisabled), r.URL)
 				return
@@ -447,15 +448,15 @@ func setRequestValidityHandler(h http.Handler) http.Handler {
 		if !globalIsTLS && (crypto.SSEC.IsRequested(r.Header) || crypto.SSECopy.IsRequested(r.Header)) {
 			if r.Method == http.MethodHead {
 				if ok {
-					tc.funcName = "handler.ValidRequest"
-					tc.responseRecorder.LogErrBody = false
+					tc.FuncName = "handler.ValidRequest"
+					tc.ResponseRecorder.LogErrBody = false
 				}
 
 				writeErrorResponseHeadersOnly(w, errorCodes.ToAPIErr(ErrInsecureSSECustomerRequest))
 			} else {
 				if ok {
-					tc.funcName = "handler.ValidRequest"
-					tc.responseRecorder.LogErrBody = true
+					tc.FuncName = "handler.ValidRequest"
+					tc.ResponseRecorder.LogErrBody = true
 				}
 
 				writeErrorResponse(r.Context(), w, errorCodes.ToAPIErr(ErrInsecureSSECustomerRequest), r.URL)

--- a/cmd/http-tracer.go
+++ b/cmd/http-tracer.go
@@ -31,6 +31,7 @@ import (
 	"github.com/minio/madmin-go"
 	"github.com/minio/minio/internal/handlers"
 	xhttp "github.com/minio/minio/internal/http"
+	"github.com/minio/minio/internal/mcontext"
 )
 
 var ldapPwdRegex = regexp.MustCompile("(^.*?)LDAPPassword=([^&]*?)(&(.*?))?$")
@@ -62,18 +63,6 @@ func getOpName(name string) (op string) {
 	return op
 }
 
-type contextTraceReqType string
-
-const contextTraceReqKey = contextTraceReqType("request-trace-info")
-
-// Hold related tracing data of a http request, any handler
-// can modify this struct to modify the trace information .
-type traceCtxt struct {
-	requestRecorder  *xhttp.RequestRecorder
-	responseRecorder *xhttp.ResponseRecorder
-	funcName         string
-}
-
 // If trace is enabled, execute the request if it is traced by other handlers
 // otherwise, generate a trace event with request information but no response.
 func httpTracer(h http.Handler) http.Handler {
@@ -84,16 +73,19 @@ func httpTracer(h http.Handler) http.Handler {
 		}
 
 		// Create tracing data structure and associate it to the request context
-		tc := traceCtxt{}
-		ctx := context.WithValue(r.Context(), contextTraceReqKey, &tc)
+		tc := mcontext.TraceCtxt{
+			AmzReqID: r.Header.Get(xhttp.AmzRequestID),
+		}
+
+		ctx := context.WithValue(r.Context(), mcontext.ContextTraceKey, &tc)
 		r = r.WithContext(ctx)
 
 		// Setup a http request and response body recorder
 		reqRecorder := &xhttp.RequestRecorder{Reader: r.Body}
 		respRecorder := xhttp.NewResponseRecorder(w)
 
-		tc.requestRecorder = reqRecorder
-		tc.responseRecorder = respRecorder
+		tc.RequestRecorder = reqRecorder
+		tc.ResponseRecorder = respRecorder
 
 		// Execute call.
 		r.Body = reqRecorder
@@ -103,7 +95,7 @@ func httpTracer(h http.Handler) http.Handler {
 		reqEndTime := time.Now().UTC()
 
 		tt := madmin.TraceInternal
-		if strings.HasPrefix(tc.funcName, "s3.") {
+		if strings.HasPrefix(tc.FuncName, "s3.") {
 			tt = madmin.TraceS3
 		}
 		// No need to continue if no subscribers for actual type...
@@ -141,7 +133,7 @@ func httpTracer(h http.Handler) http.Handler {
 		}
 
 		// Calculate function name
-		funcName := tc.funcName
+		funcName := tc.FuncName
 		if funcName == "" {
 			funcName = "<unknown>"
 		}
@@ -185,17 +177,17 @@ func httpTracer(h http.Handler) http.Handler {
 
 func httpTrace(f http.HandlerFunc, logBody bool) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
-		tc, ok := r.Context().Value(contextTraceReqKey).(*traceCtxt)
+		tc, ok := r.Context().Value(mcontext.ContextTraceKey).(*mcontext.TraceCtxt)
 		if !ok {
 			// Tracing is not enabled for this request
 			f.ServeHTTP(w, r)
 			return
 		}
 
-		tc.funcName = getOpName(runtime.FuncForPC(reflect.ValueOf(f).Pointer()).Name())
-		tc.requestRecorder.LogBody = logBody
-		tc.responseRecorder.LogAllBody = logBody
-		tc.responseRecorder.LogErrBody = true
+		tc.FuncName = getOpName(runtime.FuncForPC(reflect.ValueOf(f).Pointer()).Name())
+		tc.RequestRecorder.LogBody = logBody
+		tc.ResponseRecorder.LogAllBody = logBody
+		tc.ResponseRecorder.LogErrBody = true
 
 		f.ServeHTTP(w, r)
 	}

--- a/cmd/metrics-v2.go
+++ b/cmd/metrics-v2.go
@@ -31,6 +31,7 @@ import (
 	"github.com/minio/madmin-go"
 	"github.com/minio/minio/internal/bucket/lifecycle"
 	"github.com/minio/minio/internal/logger"
+	"github.com/minio/minio/internal/mcontext"
 	"github.com/minio/minio/internal/rest"
 	"github.com/prometheus/client_golang/prometheus"
 	dto "github.com/prometheus/client_model/go"
@@ -2395,10 +2396,10 @@ func metricsServerHandler() http.Handler {
 	}
 
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		tc, ok := r.Context().Value(contextTraceReqKey).(*traceCtxt)
+		tc, ok := r.Context().Value(mcontext.ContextTraceKey).(*mcontext.TraceCtxt)
 		if ok {
-			tc.funcName = "handler.MetricsCluster"
-			tc.responseRecorder.LogErrBody = true
+			tc.FuncName = "handler.MetricsCluster"
+			tc.ResponseRecorder.LogErrBody = true
 		}
 
 		mfs, err := gatherers.Gather()
@@ -2442,10 +2443,10 @@ func metricsNodeHandler() http.Handler {
 		registry,
 	}
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		tc, ok := r.Context().Value(contextTraceReqKey).(*traceCtxt)
+		tc, ok := r.Context().Value(mcontext.ContextTraceKey).(*mcontext.TraceCtxt)
 		if ok {
-			tc.funcName = "handler.MetricsNode"
-			tc.responseRecorder.LogErrBody = true
+			tc.FuncName = "handler.MetricsNode"
+			tc.ResponseRecorder.LogErrBody = true
 		}
 
 		mfs, err := gatherers.Gather()

--- a/cmd/metrics.go
+++ b/cmd/metrics.go
@@ -23,6 +23,7 @@ import (
 	"time"
 
 	"github.com/minio/minio/internal/logger"
+	"github.com/minio/minio/internal/mcontext"
 	iampolicy "github.com/minio/pkg/iam/policy"
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/common/expfmt"
@@ -569,10 +570,10 @@ func metricsHandler() http.Handler {
 	}
 
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		tc, ok := r.Context().Value(contextTraceReqKey).(*traceCtxt)
+		tc, ok := r.Context().Value(mcontext.ContextTraceKey).(*mcontext.TraceCtxt)
 		if ok {
-			tc.funcName = "handler.MetricsLegacy"
-			tc.responseRecorder.LogErrBody = true
+			tc.FuncName = "handler.MetricsLegacy"
+			tc.ResponseRecorder.LogErrBody = true
 		}
 
 		mfs, err := gatherers.Gather()
@@ -602,13 +603,13 @@ func metricsHandler() http.Handler {
 // AuthMiddleware checks if the bearer token is valid and authorized.
 func AuthMiddleware(h http.Handler) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		tc, ok := r.Context().Value(contextTraceReqKey).(*traceCtxt)
+		tc, ok := r.Context().Value(mcontext.ContextTraceKey).(*mcontext.TraceCtxt)
 
 		claims, groups, owner, authErr := metricsRequestAuthenticate(r)
 		if authErr != nil || !claims.VerifyIssuer("prometheus", true) {
 			if ok {
-				tc.funcName = "handler.MetricsAuth"
-				tc.responseRecorder.LogErrBody = true
+				tc.FuncName = "handler.MetricsAuth"
+				tc.ResponseRecorder.LogErrBody = true
 			}
 
 			writeErrorResponseJSON(r.Context(), w, toAdminAPIErr(r.Context(), errAuthentication), r.URL)
@@ -624,8 +625,8 @@ func AuthMiddleware(h http.Handler) http.Handler {
 			Claims:          claims.Map(),
 		}) {
 			if ok {
-				tc.funcName = "handler.MetricsAuth"
-				tc.responseRecorder.LogErrBody = true
+				tc.FuncName = "handler.MetricsAuth"
+				tc.ResponseRecorder.LogErrBody = true
 			}
 
 			writeErrorResponseJSON(r.Context(), w, toAdminAPIErr(r.Context(), errAuthentication), r.URL)

--- a/cmd/utils.go
+++ b/cmd/utils.go
@@ -59,6 +59,7 @@ import (
 	ioutilx "github.com/minio/minio/internal/ioutil"
 	"github.com/minio/minio/internal/logger"
 	"github.com/minio/minio/internal/logger/message/audit"
+	"github.com/minio/minio/internal/mcontext"
 	"github.com/minio/pkg/certs"
 	"github.com/minio/pkg/env"
 	xnet "github.com/minio/pkg/net"
@@ -915,12 +916,14 @@ func updateReqContext(ctx context.Context, objects ...ObjectV) context.Context {
 
 // Returns context with ReqInfo details set in the context.
 func newContext(r *http.Request, w http.ResponseWriter, api string) context.Context {
+	reqID := w.Header().Get(xhttp.AmzRequestID)
+
 	vars := mux.Vars(r)
 	bucket := vars["bucket"]
 	object := likelyUnescapeGeneric(vars["object"], url.PathUnescape)
 	reqInfo := &logger.ReqInfo{
 		DeploymentID: globalDeploymentID,
-		RequestID:    w.Header().Get(xhttp.AmzRequestID),
+		RequestID:    reqID,
 		RemoteHost:   handlers.GetSourceIP(r),
 		Host:         getHostName(r),
 		UserAgent:    r.UserAgent(),
@@ -929,7 +932,15 @@ func newContext(r *http.Request, w http.ResponseWriter, api string) context.Cont
 		ObjectName:   object,
 		VersionID:    strings.TrimSpace(r.Form.Get(xhttp.VersionID)),
 	}
-	return logger.SetReqInfo(r.Context(), reqInfo)
+
+	ctx := context.WithValue(r.Context(),
+		mcontext.ContextTraceKey,
+		&mcontext.TraceCtxt{
+			AmzReqID: reqID,
+		},
+	)
+
+	return logger.SetReqInfo(ctx, reqInfo)
 }
 
 // Used for registering with rest handlers (have a look at registerStorageRESTHandlers for usage example)

--- a/internal/mcontext/ctxt.go
+++ b/internal/mcontext/ctxt.go
@@ -1,0 +1,40 @@
+// Copyright (c) 2015-2022 MinIO, Inc.
+//
+// This file is part of MinIO Object Storage stack
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package mcontext
+
+// Share a common context information between different
+// packages in github.com/minio/minio
+
+import (
+	xhttp "github.com/minio/minio/internal/http"
+)
+
+// ContextTraceType represents the type of golang Context key
+type ContextTraceType string
+
+// ContextTraceKey is the key of TraceCtxt saved in a Golang context
+const ContextTraceKey = ContextTraceType("ctx-trace-info")
+
+// TraceCtxt holds related tracing data of a http request.
+type TraceCtxt struct {
+	RequestRecorder  *xhttp.RequestRecorder
+	ResponseRecorder *xhttp.ResponseRecorder
+
+	FuncName string
+	AmzReqID string
+}

--- a/internal/rest/client.go
+++ b/internal/rest/client.go
@@ -34,6 +34,7 @@ import (
 
 	xhttp "github.com/minio/minio/internal/http"
 	"github.com/minio/minio/internal/logger"
+	"github.com/minio/minio/internal/mcontext"
 	xnet "github.com/minio/pkg/net"
 )
 
@@ -192,6 +193,10 @@ func (c *Client) newRequest(ctx context.Context, u *url.URL, body io.Reader) (*h
 	req.Header.Set("X-Minio-Time", time.Now().UTC().Format(time.RFC3339))
 	if body != nil {
 		req.Header.Set("Expect", "100-continue")
+	}
+
+	if tc, ok := ctx.Value(mcontext.ContextTraceKey).(*mcontext.TraceCtxt); ok {
+		req.Header.Set(xhttp.AmzRequestID, tc.AmzReqID)
 	}
 
 	return req, nil

